File: src/test/java/com/kotovdv/just/messenger/repository/UsersTest.java
package com.kotovdv.just.messenger.repository;

import com.github.springtestdbunit.annotation.DatabaseSetup;
import com.github.springtestdbunit.annotation.ExpectedDatabase;
import com.kotovdv.just.messenger.entity.User;
import com.kotovdv.just.messenger.repository.configuration.RepositoryTest;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import static com.github.springtestdbunit.assertion.DatabaseAssertionMode.NON_STRICT;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author Dmitriy Kotov
 */
public class UsersTest extends RepositoryTest {

    @Autowired
    private Users users;

    @Autowired
    private TestEntityManager testEntityManager;

    @Test
    @ExpectedDatabase(value = "classpath:users/add/after.xml", assertionMode = NON_STRICT)
    public void testUserAdd() throws Exception {
        User user = new User("John", "Doe");
        users.save(user);
    }

    @Test
    @DatabaseSetup(value = "classpath:users/get/before.xml")
    public void testUserGet() {
        User user = users.findOne(1L);

        assertThat(user.getId()).isEqualTo(1L);
        assertThat(user.getFirstName()).isEqualTo("John");
        assertThat(user.getLastName()).isEqualTo("Doe");
    }

    @Test
    @DatabaseSetup("classpath:users/count/before.xml")
    public void testUserCount() {
        assertThat(users.count()).isEqualTo(3);
    }

    @Test
    @DatabaseSetup("classpath:users/remove_entity/before.xml")
    @ExpectedDatabase(value = "classpath:users/remove_entity/after.xml", assertionMode = NON_STRICT)
    public void testRemoveByEntityUser() throws Exception {
        User user = testEntityManager.find(User.class, 1L);

        users.delete(user);
        testEntityManager.flush();
    }

    @Test
    @DatabaseSetup("classpath:users/remove_id/before.xml")
    @ExpectedDatabase(value = "classpath:users/remove_id/after.xml", assertionMode = NON_STRICT)
    public void testRemoveByIdUser() throws Exception {
        users.delete(1L);
        testEntityManager.flush();
    }
}

File: src/test/java/com/kotovdv/just/messenger/repository/UsersTest.kt
package com.kotovdv.just.messenger.repository

import com.github.springtestdbunit.annotation.DatabaseSetup
import com.github.springtestdbunit.annotation.ExpectedDatabase
import com.github.springtestdbunit.assertion.DatabaseAssertionMode.NON_STRICT
import com.kotovdv.just.messenger.model.entity.User
import com.kotovdv.just.messenger.repository.assertion.ExpectedQueriesCount
import com.kotovdv.just.messenger.repository.common.RepositoryTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Import
import java.time.LocalDateTime.of

@Import(Users::class)
class UsersTest : RepositoryTest() {

    @Autowired
    lateinit var users: Users

    private val CREATION_DATE = of(1970, 1, 1, 12, 0)

    @Test
    @ExpectedDatabase(value = "classpath:users/add/after.xml", assertionMode = NON_STRICT)
    @ExpectedQueriesCount(1)
    fun testUserAdd() {
        val user = User(null, "John", "Doe", CREATION_DATE)
        users.addOrUpdate(user)
    }

    @Test
    @DatabaseSetup(value = "classpath:users/get/before.xml")
    @ExpectedQueriesCount(1)
    fun testUserGet() {
        val user = users.get(1L) ?: throw AssertionError()

        assertThat(user.id).isEqualTo(1L)
        assertThat(user.firstName).isEqualTo("John")
        assertThat(user.lastName).isEqualTo("Doe")
        assertThat(user.creationDate).isEqualTo(CREATION_DATE)
    }

    @Test
    @DatabaseSetup("classpath:users/size/before.xml")
    @ExpectedQueriesCount(1)
    fun testUserSize() {
        assertThat(users.size()).isEqualTo(3)
    }

    @Test
    @DatabaseSetup("classpath:users/remove/before.xml")
    @ExpectedDatabase(value = "classpath:users/remove/after.xml", assertionMode = NON_STRICT)
    @ExpectedQueriesCount(4)
    fun testUserRemove() {
        users.remove(1L)
        testEntityManager.flush()
    }

    @Test
    @DatabaseSetup("classpath:users/chat_list/before.xml")
    @ExpectedQueriesCount(2)
    fun testUserChatList() {
        val user = users.get(1L) ?: throw AssertionError()

        assertThat(user.chats.size).isEqualTo(3)
    }
}
